options/options_helper.cc

```diff
@@ -1102,41 +1001,14 @@ static Status ParseDBOption(const ConfigOptions& config_options,
   const std::string& value = config_options.input_strings_escaped
                                  ? UnescapeOptionString(org_value)
                                  : org_value;
-  try {
-    if (name == "rate_limiter_bytes_per_sec") {
-      new_options->rate_limiter.reset(
-          NewGenericRateLimiter(static_cast<int64_t>(ParseUint64(value))));
-    } else if (name == kNameEnv) {
-      // Currently `Env` can be deserialized from object registry only.
-      Env* env = new_options->env;
-      Status status = Env::LoadEnv(value, &env);
-      // Only support static env for now.
-      if (status.ok()) {
-        new_options->env = env;
-      }
-    } else {
-      auto iter = db_options_type_info.find(name);
-      if (iter == db_options_type_info.end()) {
-        return Status::InvalidArgument("Unrecognized option DBOptions:", name);
-      }
-      const auto& opt_info = iter->second;
-      if (opt_info.IsDeprecated() ||
-          ParseOptionHelper(
-              reinterpret_cast<char*>(new_options) + opt_info.offset,
-              opt_info.type, value)) {
-        return Status::OK();
-      } else if (opt_info.IsByName()) {
-        return Status::NotSupported("Deserializing the specified DB option " +
-                                    name + " is not supported");
-      } else {
-        return Status::InvalidArgument(
-            "Unable to parse the specified DB option " + name);
-      }
-    }
-  } catch (const std::exception&) {
-    return Status::InvalidArgument("Unable to parse DBOptions:", name);
+  auto iter = db_options_type_info.find(name);
+  if (iter == db_options_type_info.end()) {
+    return Status::InvalidArgument("Unrecognized option DBOptions:", name);
```

Comment:
Would be better to ensure the error messanges are the same. For CF cases and also DB cases. After the change, the error message will be like "Deserializing the option xxxx is not supported" and "Error parsing xxxx". But the original one will be "Deserializing the specified DB option xxxx is not supported" or "Deserializing the specified CF option xxxx is not supported". Similar for "Unable to parse the specified DB option" and others

Reply:
- If the option is not found (iter==end), the output is exactly the same
- If we caught an exception, the status is the same (InvalidArgument) but now also contains information about the exception.
- If we failed to parse but it was a "Name" option, the status is the same (NotSupported) but no longer contains "DBOption" in the text.  This allows the unification of the code 
- If the value could not be parsed, the status is the same but now says "Error parsing"  rather than  "unable to parse specified DBOption".

If you look around the code base, this nearly identical code was in several places and each treated the errors slightly differently and printed slightly different messages.  This code is a first pass at an attempt to unify the messages and errors that are returned.  In the end, I would like to see NotSupported, NotFound, and InvalidArgument mean specific things for all options:
- NotSupported: It is a valid option but we do not know how to parse the value
- InvalidArgument: It is a valid option name but we cannot parse the value for some reason
- NotFound: It is not a valid option name
Unifying these errors will make it easier to document them, expand the use/scope of options, and allow the caller to know what to expect.

Reply:
> * If the option is not found (iter==end), the output is exactly the same
> * If we caught an exception, the status is the same (InvalidArgument) but now also contains information about the exception.
> * If we failed to parse but it was a "Name" option, the status is the same (NotSupported) but no longer contains "DBOption" in the text.  This allows the unification of the code
> * If the value could not be parsed, the status is the same but now says "Error parsing"  rather than  "unable to parse specified DBOption".
> 
> If you look around the code base, this nearly identical code was in several places and each treated the errors slightly differently and printed slightly different messages. This code is a first pass at an attempt to unify the messages and errors that are returned. In the end, I would like to see NotSupported, NotFound, and InvalidArgument mean specific things for all options:
> 
> * NotSupported: It is a valid option but we do not know how to parse the value
> * InvalidArgument: It is a valid option name but we cannot parse the value for some reason
> * NotFound: It is not a valid option name
>   Unifying these errors will make it easier to document them, expand the use/scope of options, and allow the caller to know what to expect.

It is clear for me now. It would be better to add these explanation of error message as the comments above ParseOption.

Reply:
I added the error status codes to the ParseOption and GetXXXFromMap methods.